testing_utilities/approximate_quantity_body.hpp

```diff
@@ -0,0 +1,161 @@
+﻿#pragma once
+
+#include <optional>
+
+#include "testing_utilities/approximate_quantity.hpp"
+
+namespace principia {
+namespace testing_utilities {
+namespace internal_approximate_quantity {
+
+template<typename Dimensions>
+Quantity<Dimensions> ApproximateQuantity<Quantity<Dimensions>>::min() const {
+  return min_multiplier_ * unit_;
+}
+
+template<typename Dimensions>
+Quantity<Dimensions> ApproximateQuantity<Quantity<Dimensions>>::max() const {
+  return max_multiplier_ * unit_;
+}
+
+template<typename Dimensions>
+std::string ApproximateQuantity<Quantity<Dimensions>>::DebugString() const {
+  return "[" + quantities::DebugString(min_multiplier_) + ", " +
+         quantities::DebugString(max_multiplier_) + "] * " +
+         quantities::DebugString(unit_);
+}
+
+template<typename Dimensions>
+ApproximateQuantity<Quantity<Dimensions>>::ApproximateQuantity(
+    std::string const& representation,
+    double const min_multiplier,
+    double const max_multiplier,
+    Quantity<Dimensions> const& unit)
+    : representation_(representation),
+      min_multiplier_(min_multiplier),
+      max_multiplier_(max_multiplier),
+      unit_(unit) {}
+
+ApproximateQuantity<double> ApproximateQuantity<double>::Parse(
+    char const* const representation,
+    int const ulp) {
+  std::string error_representation(representation);
+  std::optional<int> last_digit_index;
+  bool const is_hexadecimal =
+      error_representation.size() >= 2 &&
+      error_representation[0] == '0' &&
+      (error_representation[1] == 'x' || error_representation[1] == 'X');
+  for (int i = 0; i < error_representation.size(); ++i) {
```

Comment:
I think that this would be more readable if we were building `error_representation` (by looping on `representation` and appending) than rewriting it with 0s; either way, add a comment describing what we are doing.

Reply:
It would not be more readable because I would need more state variables to copy the exponent without patching it.  Added a comment.